File: marketsim/egta/utils/simplex_operations.py
import numpy as np
import torch
from scipy.special import comb
from itertools import combinations_with_replacement
from egta.utils.log_multimodal import logmultinomial, logmultinomial_torch
'''
original code from Bryce here: 
https://github.com/Davidson-Game-Theory-Research/gameanalysis.jl/blob/master/SimplexOperations.jl
'''
def uniform_mixture(num_actions):
    '''
    This returns the cener point of the simplex
    '''

    return torch.ones(num_actions) / num_actions

def random_mixture(num_actions, num_mixtures, alpha=1.0, device="cpu"):
    '''
    This returns random points on the simplex
    
    Parameters:
    num_actions : int
        Number of dimensions in the simplex
    num_mixtures : int
        Number of random points to generate
    alpha : float or list
        Concentration parameter for Dirichlet distribution
    device : str
        Device to place the tensor on ("cpu" or "cuda")
    Returns: 
    torch.Tensor : Tensor of shape (num_actions, num_mixtures) with random points on the simplex
    '''
    if isinstance(alpha, (list, tuple, np.ndarray)):
        #i f alpha is an array, use it directly as concentration parameters
        alpha_tensor = torch.tensor(alpha, device=device)
        return torch.distributions.Dirichlet(alpha_tensor).sample((num_mixtures,)).T
    else:
        #use uniform concentration parameter
        alpha_tensor = torch.ones(num_actions, device=device) * alpha
        return torch.distributions.Dirichlet(alpha_tensor).sample((num_mixtures,)).T
    
def sample_profile(num_players, mixture, device="cpu"):
    '''
    This samples a profile from a given mixture
    '''
    mixture = torch.as_tensor(mixture, device=device)
    return torch.distributions.Multinomial(num_players, probs=mixture).sample()

def sample_profiles(num_players, mixtures, device="cpu"):
    '''
    Sample profiles from multiple mixtures
    Parameters:
    num_players : int
        Number of players in the game
    mixtures : torch.Tensor
        Matrix where each column is a mixture
    device : str
        Device to place the tensor on
    Returns:
    torch.Tensor : Matrix where each column is a sampled profile
    '''
    mixtures = torch.as_tensor(mixtures, device=device)
    profiles = torch.zeros_like(mixtures)
    for m in range(mixtures.shape[1]):
        profiles[:, m] = sample_profile(num_players, mixtures[:, m], device)
    return profiles

def num_profiles(num_players, num_actions):
    '''
    calculates the number of distinct profiles in a symmetric game
    
    num_players : int
        Number of players in the game
    num_actions : int
        Number of available actions

    Returns:
    int : Number of distinct profiles
    '''
    return comb(num_players + num_actions - 1, num_actions - 1, exact=True)

def simplex_projection(y, device="cpu"):

    '''
    This projects a vector onto the probability simplex
    This follows what Bryce did in his julia code

    returns a projected pytorch tensor 
    '''
    y = torch.as_tensor(y, device=device)
    if len(y.shape) == 1: #if we have a single vector 
        u = torch.sort(y, descending=True)[0]
        cumulative_sum = torch.cumsum(u, dim=0)
        rho = torch.nonzero(u > (cumulative_sum - 1)/ (torch.arange(1, len(y) + 1, device=device)))[-1]
        theta = (cumulative_sum[rho] - 1) / (rho + 1)
        return torch.max(torch.zeros_like(y), y - theta)
    else: #matrix case, here we project to each column
        result = torch.zeros_like(y)
        for i in range(y.shape[1]):
            result[:,i] = simplex_projection(y[:,i], device=device) #recurse on each column
        return result

def simplex_normalize(d, epsilon=1e-10):
    '''
    normalize to the probability simplex
    returns normalized pytorch tensor
    '''
    d = torch.as_tensor(d)
    d = torch.clamp(d, min=epsilon)
    return d / torch.sum(d)


def profile_rankings(profile):
    '''
    ranking algorithm from Bryce's paper with replacement 
    used to determine the index of the profile in the configuration table
    '''
    profile = torch.as_tensor(profile)
    num_actions = len(profile)
    num_opponents = torch.sum(profile) 
    preceding_profiles = 0

    for a in range(num_actions):
        num_opponents -= profile[a] 
        stars = num_opponents - 1
        bars = num_actions + 1 - a
        if stars >= 0 and bars > 0:
            preceding_profs += torch.special.binom(stars + bars - 1, stars)
    
    return int(preceding_profs + 1)

def mixture_grid(num_actions, points_per_dim, device="cpu"):

    '''
    creates a grid equally spaced points throughout the simplex
    Parameters:

    num_actions : int
        Number of dimensions in the simplex
    points_per_dim : int
        Number of points along each dimension
    device : str
        Device to place the tensor on

    Returns:
    torch.Tensor : Matrix where each column is a point on the grid
    '''
    num_mixtures = comb(num_actions - 1, points_per_dim - 1, exact = True)
    mixtures = torch.zeros((num_actions, num_mixtures), device=device)

    #generate points on unit simplex using stars and bars algorithm:
    # https://en.wikipedia.org/wiki/Stars_and_bars_(combinatorics)
    for m, config in enumerate(combinations_with_replacement(range(1, num_actions + 1), points_per_dim - 1)):
        #count occurrences of each action in the configuration
        counts = torch.zeros(num_actions, device=device)
        for c in config:
            counts[c - 1] += 1
        
        #normalize to create a point on the simplex
        mix = counts / (points_per_dim - 1)
        mixtures[:, m] = mix
    
    return mixtures

def simplex_projection_sum(y, simplex_sum=1.0, device="cpu"):
    '''
    projects onto a smaller sub-simplex where the dimensions add to simplex_sum
    
    Parameters:
    y : torch.Tensor
        Vector or matrix to project
    simplex_sum : float
        Target sum for the simplex
    device : str
        Device to place the tensor on

    Returns:
    torch.Tensor: projected vector/matrix
    '''
    y = torch.as_tensor(y, device = device)

    if len(y.shape) == 1: #just a single vector
        D = len(y)
        u = torch.sort(y, descending=True)[0]

        cumulative_sum = torch.cumsum(u, dim=0)
        lambda_vector = (simplex_sum - cumulative_sum) / torch.arange(1, D+1, device=device)

        valid_indicies = torch.nonzero(u + lambda_vector > 0)
        if len(valid_indicies) == 0:
            lambda_value = lambda_vector[-1]

        else:
            lambda_value = lambda_vector[valid_indicies[-1]]
        return torch.max(torch.zeros_like(y), y + lambda_value) + (1 - simplex_sum) / D
    
    else: #matrix case
        D = y.shape[0]
        u = torch.sort(y, dim=0, descending=True)[0]
        cumulative_sum = torch.cumsum(u, dim=0)
        lambda_vector = (simplex_sum - cumulative_sum) / torch.arange(1, D+1, 
                                                                   device=device).reshape(-1, 1)
        
        lambda_values = torch.zeros(y.shape[1], device=device)

        for i in range(y.shape[1]):
            valid_indices = torch.nonzero(u[:, i] + lambda_vector[:, i] > 0)

            if len(valid_indices) == 0:
                lambda_values[i] = lambda_vector[-1, i]
            else:
                lambda_values[i] = lambda_vector[valid_indices[-1], i]
        return torch.max(torch.zeros_like(y), y + lambda_values.reshape(1, -1)) + (1 - simplex_sum) / D
    

def simplex_normalize(d, epsilon=1e-10):
    '''
    normalize to the probability simplex
    returns normalized pytorch tensor
    '''
    d = torch.as_tensor(d)
    d = torch.clamp(d, min=epsilon)
    return d / torch.sum(d, dim=0)



def filter_unique(mixtures, max_diff=1e-2, device="cpu"):
    '''
    filter a collection of mixed strategies of mixed strategies
    to remove approximate duplicates 
    keep the first appearance of the group

    Parameters:
    mixtures : torch.Tensor
        Matrix where each column is a mixture
    max_diff : float
        Maximum difference for two mixtures to be considered duplicates
    device : str
        Device to place the tensor on

    Returns:
    torch.Tensor : Matrix with duplicates removed
    '''

    mixtures = torch.as_tensor(mixtures, device=device)
    if mixtures.shape[1] == 0:
        return mixtures
    
    unique_indices = [0]  #keeps first
    
    for m in range(1, mixtures.shape[1]):
        is_unique = True
        for i in unique_indices:
            if torch.max(torch.abs(mixtures[:, i] - mixtures[:, m])) < max_diff:
                is_unique = False
                break
        if is_unique:
            unique_indices.append(m)
    
    return mixtures[:, unique_indices]

def num_payoffs(num_players, num_actions, dev=True):
    '''
    compute the number of payoffs in a symmetric games
    Parameters:
    num_players : int
        Number of players in the game
    num_actions : int
        Number of available actions
    dev : bool
        Whether to count deviation payoffs (True) or pure payoffs (False)
    Returns:
    int : Number of payoffs
    '''
    if dev:
        return int(np.exp(logmultinomial(num_players-1, num_actions-1)) * num_actions)
    else:
        return int(np.exp(logmultinomial(num_players, num_actions-1)) * num_actions)

File: bryce_code/SimplexOperations.jl
using Distributions: Dirichlet, Multinomial
using Combinatorics: multinomial, with_replacement_combinations as CwR
using StatsBase: counts

include("LogMultinomial.jl")

# center-point of the num_actions-simplex
function uniform_mixture(num_actions::Integer)
    ones(num_actions) / num_actions
end

# draws independent samples from the num_actions-simplex
function random_mixtures(num_actions::Integer, num_mixtures::Integer, α=1; GPU=false)
    if α isa Array
        distr = Dirichlet(α)
    else
        distr = Dirichlet(num_actions, α)
    end
    mixtures = rand(distr, num_mixtures)
    if GPU
        return CUDA.CuArray{Float32,2}(mixtures)
    else
        return mixtures
    end
end

# samples one random profile from a given mixture
function sample_profile(num_players::Integer, mixture::AbstractVector)
    return rand(Multinomial(num_players, mixture))
end

# samples one random profile from each of the given mixtures
function sample_profiles(num_players::Integer, mixtures::AbstractMatrix)
    profiles = Matrix{Int64}(undef, size(mixtures))
    for m in axes(mixtures,2)
        profiles[:,m] .= sample_profile(num_players, mixtures[:,m])
    end
    return profiles
end

# number of opponent-profiles in a game with P players and A actions
function num_profiles(num_players::Integer, num_actions::Integer)
    multinomial(num_players-1, num_actions-1)
end

# ranking algorithm for combinations with replacement
# used to determine a profile's index in the configuration table
function profile_ranking(profile)
    num_actions = length(profile)
    num_opponents = sum(profile)
    preceding_profs = 0
    for a in 1:num_actions
        num_opponents -= profile[a]
        stars = num_opponents - 1
        bars = num_actions + 1 - a
        preceding_profs += binomial(stars + bars - 1, stars)
    end
    return preceding_profs + 1
end

# creates a grid of equally spaced points throughout the simplex
function mixture_grid(num_actions::Integer, points_per_dim::Integer)
    num_mixtures = multinomial(num_actions-1, points_per_dim-1)
    mixtures = Array{Float64,2}(undef, num_actions, num_mixtures)
    for (m,config) in enumerate(CwR(1:num_actions, points_per_dim-1))
        mix = counts(config, 1:num_actions) ./ (points_per_dim-1)
        mixtures[:,m] = mix
    end
    return mixtures
end

# creates a grid of equally spaced points within the sub-simplex specified by corners
function mixture_grid(corners::AbstractMatrix, points_per_dim::Integer)
    grid = mixture_grid(size(corners,2), points_per_dim)
    return corners * grid
end

# identifies the largest mixture_grid with at most max_mixtures elements
function finest_grid(num_actions::Integer, max_mixtures::Integer)
    points_per_dim = 2
    num_mixtures = num_actions
    for p in 3:max_mixtures
        m = multinomial(num_actions-1, p-1)
        if m > max_mixtures
            break
        else
            num_mixtures = m
            points_per_dim = p
        end
    end
    return (points_per_dim, num_mixtures)
end

# creates a grid of equally spaced points surrounding a given midpoint
function grid_around(midpoint::AbstractVector, scale::AbstractFloat, points_per_dim::Integer)
    num_actions = size(midpoint,1)
    offsets = (mixture_grid(num_actions, points_per_dim) .- 1/3) .* scale
    midpoint = simplex_project(midpoint, 1-scale)
    return midpoint .+ offsets
end

# creates combined grids surrounding several midpoints given by the array's columns
function grid_around(midpoints::AbstractMatrix, scale::AbstractFloat, points_per_dim::Integer; prevent_overlap=false)
    num_actions = size(midpoints,1)
    num_grids = size(midpoints,2)
    mixtures_per_grid = multinomial(num_actions-1, points_per_dim-1)
    grids = Matrix{Float64}(undef, num_actions, mixtures_per_grid * num_grids)
    for g in 1:num_grids
        grids[:,(g-1)*mixtures_per_grid+1:g*mixtures_per_grid] .= grid_around(midpoints[:,g], scale, points_per_dim)
    end
    if prevent_overlap
        return unique(grids, dims=2)
    else
        return grids
    end
end

# notation follows https://arxiv.org/pdf/1309.1541.pdf
function simplex_project(y⃗::AbstractVector)
    D = length(y⃗)
    u⃗  = sort(y⃗, rev=true)
    λ⃗ = (1 .- cumsum(u⃗)) ./ (1:D)
    λ = λ⃗[findlast(u⃗ .+ λ⃗ .> 0)]
    max.(y⃗ .+ λ, 0)
end

# version that vectorizes over the columns of M
function simplex_project(M::AbstractMatrix)
    D = size(M,1)
    u⃗  = sort(M, dims=1, rev=true)
    λ⃗ = (1 .- cumsum(u⃗, dims=1)) ./ (1:D)
    λ⃗[u⃗ .+ λ⃗ .< 0] .= Inf
    λ = minimum(λ⃗, dims=1)
    max.(M .+ λ, 0)
end

# projects onto a smaller sub-simplex where the dimensions add to simplex_sum
function simplex_project(y⃗::AbstractVector, simplex_sum::Real)
    D = length(y⃗)
    u⃗  = sort(y⃗, rev=true)
    λ⃗ = (simplex_sum .- cumsum(u⃗)) ./ (1:D)
    λ = λ⃗[findlast(u⃗ .+ λ⃗ .> 0)]
    max.(y⃗ .+ λ, 0) .+ (1 - simplex_sum) ./ D
end

# both vectorized and sub-simplex
function simplex_project(M::AbstractMatrix, simplex_sum::Real)
    D = size(M,1)
    u⃗  = sort(M, dims=1, rev=true)
    λ⃗ = (simplex_sum .- cumsum(u⃗, dims=1)) ./ (1:D)
    λ⃗[u⃗ .+ λ⃗ .< 0] .= Inf
    λ = minimum(λ⃗, dims=1)
    max.(M .+ λ, 0) .+ (1 - simplex_sum) ./ D
end

# Normalize distribution(s) by flooring at zero and then dividing by the sum.
function simplex_normalize(d::AbstractVecOrMat, ϵ=eps(0.0))
    d = copy(d)
    d[d .< ϵ] .= ϵ
    d ./ sum(d, dims=1)
end

# Filter a collection of mixed strategies to remove approximate duplicates.
# Will keep the first appearance of a group, so it's a good idea to pre-sort by priority.
function filter_unique(mixtures::Matrix; max_diff=1e-2)
    if size(mixtures,2) == 0
        return mixtures
    end
    unique = Matrix{eltype(mixtures)}(undef, size(mixtures)...)
    unique[:,1] .= mixtures[:,1]
    count = 1
    for m in 2:size(mixtures,2)
        is_unique = true
        for i in 1:count
            if maximum(unique[:,i] .- mixtures[:,m]) < max_diff
                is_unique = false
                continue
            end
        end
        if is_unique
            count += 1
            unique[:,count] .= mixtures[:,m]
        end
    end
    return Array(unique[:,1:count])
end

function num_payoffs(num_players::Integer, num_actions::Integer; dev=true)
    if dev
        return exp(logmultinomial(num_players-1, num_actions-1)) * num_actions
    else
        return exp(logmultinomial(num_players, num_actions-1)) * num_actions
    end
end